pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java

```diff
@@ -206,41 +179,318 @@ private Map<String, RoutingTable> getRoutingTable(String logicalTableName, long
+    int stageParallelism = Integer.parseInt(options.getOrDefault(QueryOptionKey.STAGE_PARALLELISM, "1"));
+    if (metadata.isRequiresSingletonInstance()) {
+      // require singleton should return a single global worker ID with 0;
+      ServerInstance serverInstance = serverInstances.get(RANDOM.nextInt(serverInstances.size()));
+      metadata.setServerInstanceToWorkerIdMap(
+          Collections.singletonMap(new QueryServerInstance(serverInstance), Collections.singletonList(0)));
+      metadata.setTotalWorkerCount(1);
+    } else {
+      Map<QueryServerInstance, List<Integer>> serverInstanceToWorkerIdMap = new HashMap<>();
+      int nextWorkerId = 0;
+      for (ServerInstance serverInstance : serverInstances) {
+        List<Integer> workerIds = new ArrayList<>();
+        for (int i = 0; i < stageParallelism; i++) {
+          workerIds.add(nextWorkerId++);
+        }
+        serverInstanceToWorkerIdMap.put(new QueryServerInstance(serverInstance), workerIds);
+      }
+      metadata.setServerInstanceToWorkerIdMap(serverInstanceToWorkerIdMap);
+      metadata.setTotalWorkerCount(nextWorkerId);
+    }
+
+    for (PlanFragment child : fragment.getChildren()) {
+      assignWorkersToNonRootFragment(child, context);
+    }
   }
 
-  private Set<ServerInstance> fetchServersForIntermediateStage(Set<String> tableNames) {
-    Set<ServerInstance> serverInstances = new HashSet<>();
+  private boolean isColocatedJoin(PlanNode planNode) {
+    if (planNode instanceof JoinNode) {
+      return ((JoinNode) planNode).isColocatedJoin();
+    }
+    for (PlanNode child : planNode.getInputs()) {
+      if (isColocatedJoin(child)) {
+        return true;
+      }
+    }
+    return false;
+  }
 
-    for (String table : tableNames) {
-      String rawTableName = TableNameBuilder.extractRawTableName(table);
-      TableType tableType = TableNameBuilder.getTableTypeFromTableName(table);
-      if (tableType == null) {
-        String offlineTable = TableNameBuilder.forType(TableType.OFFLINE).tableNameWithType(rawTableName);
-        String realtimeTable = TableNameBuilder.forType(TableType.REALTIME).tableNameWithType(rawTableName);
+  private void assignWorkersForColocatedJoin(PlanFragment fragment, DispatchablePlanContext context) {
+    List<PlanFragment> children = fragment.getChildren();
+    Preconditions.checkArgument(children.size() == 2, "Expecting 2 children, find: %s", children.size());
+    PlanFragment leftFragment = children.get(0);
+    PlanFragment rightFragment = children.get(1);
+    Map<Integer, DispatchablePlanMetadata> metadataMap = context.getDispatchablePlanMetadataMap();
+    DispatchablePlanMetadata leftMetadata = metadataMap.get(leftFragment.getFragmentId());
+    Preconditions.checkArgument(isLeafPlan(leftMetadata), "Left side is not leaf");
+    DispatchablePlanMetadata rightMetadata = metadataMap.get(rightFragment.getFragmentId());
+    Preconditions.checkArgument(isLeafPlan(rightMetadata), "Right side is not leaf");
```

Comment:
follow up i think this limitation can be eliminated. 

for example A JOIN B JOIN C and all colocated. then the result of A JOIN B is also colocated when joining C

Reply:
Current implementation works only if we are directly accessing the table. In order to do multiple joins colocated, we need to find a way to gather all tables and pick the common servers from all of them. Added a TODO